### PR TITLE
[RFC1] Game engine initial design

### PR DESCRIPTION
**Pull request by Scrabouillmax**

### CODE REVIEW

File: rfc/rfc1-game-engine-initial-design.md
Comment (nockty) [inline rfc/rfc1-game-engine-initial-design.md:8]: Nit: `bidding` or `betting`? just wondering what is the most appropriate terminology.

File: rfc/rfc1-game-engine-initial-design.md
Comment (nockty) [inline rfc/rfc1-game-engine-initial-design.md:48]: Why do we need a timestamp for an event?

File: rfc/rfc1-game-engine-initial-design.md
Comment (nockty) [inline rfc/rfc1-game-engine-initial-design.md:51]: That's great!

**Review by nockty — COMMENTED** [review]
Great work! :tada: 

Some comments regarding the diagram:
Important:
- `BiddingGameState.getAuthorisedBiddingsForCurrentPlayer` (I hope we'll find a shorter name) should likely return an array of integers, not an array of contracts (the suit isn't necessary).
- `Contract` should also have a `Player` or `Team` field. (Should we have a `Team` class?)
- the `trumpSuit` in `MainGameState` is unnecessary, as `getTrumpSuit` can return `contract.suit`. Note that I think we shouldn't bother with "all trumps" and "no trumps" contracts for now.
- I think the `Trick` should associate players with the card they played. Is the `username` field there to know who won the trick?
- `PlayerBadeEvent` should also handle the case when a player _coinched_ or _surcoinched_. Maybe by adding fields `bool coinche` and `bool surcoinche`? If we decide to use these fields, they should also be in the `BiddingGameState`, which should also store the _coinche_ or _surcoinche_ as well as the highest bidding.

Less important:
- I think we'll see during the implementation if we need abstract classes or interfaces. We should use the one which makes the most sense but it's difficult to see for now.
- regarding methods that are exposed: let's be conservative. For instance, I'm not sure `State.getCards` is very useful. I think we'll see what's really necessary during the implementation anyway.

Let me know what you think about this.

File: rfc/rfc1-game-engine-initial-design.md
Comment (Scrabouillmax) [reply rfc/rfc1-game-engine-initial-design.md:8]: To me betting is more when there is money at stake, typically in poker we would say 'a bet' ;)
I think both still make sense though, your call ;)

File: rfc/rfc1-game-engine-initial-design.md
Comment (Scrabouillmax) [reply rfc/rfc1-game-engine-initial-design.md:48]: - An event corresponds to something (for now a player action only, but could very well be a disconnection event, a player sending a message in a chat, a commit in git...) that happened once and for all, and can't be changed. For this reason, it makes sense to associate an event with a timestamp (it's usually the case in event-driven systems).
- In our example, we can use this timestamp to know when a player played a card to compute how much time is remaining for the following player to play his card.

**Comment by Scrabouillmax** [comment]
> BiddingGameState.getAuthorisedBiddingsForCurrentPlayer (I hope we'll find a shorter name) should likely return an array of integers, not an array of contracts (the suit isn't necessary).
>
-> I updated it and the return value. I changed the name to getAuthorisedActions, same as for the getAuthorisedPlaysForCurrentPlayer.

> Contract should also have a Player or Team field. (Should we have a Team class?)
>
-> Good catch, I added a playerId to the Contract class. For the teams, we can maybe do without for now, as we can deduce them directly from the player position, what do you think?

> the trumpSuit in MainGameState is unnecessary, as getTrumpSuit can return contract.suit. Note that I think we shouldn't bother with "all trumps" and "no trumps" contracts for now.
>
-> I agree, I removed it.

> I think the Trick should associate players with the card they played. Is the username field there to know who won the trick?
>
-> It's already the case because the trick is an array of CardPlayedEvent, each of which contains both the card and the player. I renamed the field to `cardsPlayedEvents` for more clarity. Good point for the username, I think it was a copy-paste error. I replaced `username` by `winnerId`. 

> PlayerBadeEvent should also handle the case when a player coinched or surcoinched. Maybe by adding fields bool coinche and bool surcoinche? If we decide to use these fields, they should also be in the BiddingGameState, which should also store the coinche or surcoinche as well as the highest bidding.
>
-> Good point, I added the booleans both to the event and the BiddingGameState.

> I think we'll see during the implementation if we need abstract classes or interfaces. We should use the one which makes the most sense but it's difficult to see for now.
>
-> I agree.

> regarding methods that are exposed: let's be conservative. For instance, I'm not sure State.getCards is very useful. I think we'll see what's really necessary during the implementation anyway.
>
-> I took exactly the methods that we listed when we thought about what the client would need. If any of those is not used, we will remove them for sure.

Nice comments, it's getting there 👍

File: rfc/rfc1-game-engine-initial-design.md
Comment (nockty) [reply rfc/rfc1-game-engine-initial-design.md:8]: Yeah, let's keep `bidding`

File: rfc/rfc1-game-engine-initial-design.md
Comment (nockty) [reply rfc/rfc1-game-engine-initial-design.md:48]: I'm not sure we should mix apples and oranges and have events for both game moves and chat messages or other things not related to the actual game. It's likely better if we decouple the game engine from everything else, so that we can easily change the different parts of the system in the future.

It could indeed be useful for the timer. Let's have timestamps for now, as it'll be easier to remove them than add them in the future.

**Review by nockty — APPROVED** [review]
> -> I took exactly the methods that we listed when we thought about what the client would need. If any of those is not used, we will remove them for sure.

I see. Maybe it'll make more sense to have some kind of `toJson` method to get the full state of the game at once instead of calling many different methods. It'll be more clear during the implementation anyway so we can leave this for now. 👍

The RFC looks good to me, thank you for working on it! Can you include the diagram as a png file in the markdown before merging? (It's better than having an external link I think.) You can add the file in the folder and reference it with `name`.

**Pull request merged**
